crates/ruff_source_file/src/line_ranges.rs

```diff
@@ -295,6 +295,40 @@ pub trait LineRanges {
     /// ## Panics
     /// If the start or end of `range` is out of bounds.
     fn full_lines_str(&self, range: TextRange) -> &str;
+
+    /// Returns the zero-based index of the line containing `offset`.
+    ///
+    /// ## Examples
+    ///
+    /// ```
+    /// # use ruff_text_size::{Ranged, TextRange, TextSize};
+    /// # use ruff_source_file::LineRanges;
+    ///
+    /// let text = "First line\nsecond line\r\nthird line";
+    ///
+    /// assert_eq!(text.count_lines_until(TextSize::from(5)), 0);
+    /// assert_eq!(text.count_lines_until(TextSize::from(23)), 1);
+    /// assert_eq!(text.count_lines_until(TextSize::from(24)), 2);
+    /// ```
+    ///
+    /// ## Panics
+    /// If `offset` is out of bounds.
+    fn count_lines_until(&self, offset: TextSize) -> u32 {
+        let mut count = 0;
+        let mut last_line_end = TextSize::default();
+
+        loop {
+            last_line_end = self.full_line_end(last_line_end);
+
+            if last_line_end <= offset {
+                count += 1;
+            } else {
+                break;
+            }
+        }
+
+        count
+    }
```

Comment:
Nit: i'd change the method to take a text range instead. It makes it more flexible and it also avoids that the function never terminates if `offset` == `len(text)`
```suggestion
    /// Returns the zero-based index of the line containing `offset`.
    ///
    /// ## Examples
    ///
    /// ```
    /// # use ruff_text_size::{Ranged, TextRange, TextSize};
    /// # use ruff_source_file::LineRanges;
    ///
    /// let text = "First line\nsecond line\r\nthird line";
    ///
    /// assert_eq!(text.count_lines_until(TextSize::from(5)), 0);
    /// assert_eq!(text.count_lines_until(TextSize::from(23)), 1);
    /// assert_eq!(text.count_lines_until(TextSize::from(24)), 2);
    /// ```
    ///
    /// ## Panics
    /// If `offset` is out of bounds.
    fn count_lines(&self, range: TextRange) -> u32 {
        let mut count = 0;
        let mut offset = range.start();

				loop {
					let line_end = self.full_line_end(offset);
					if line_end < range.end() {
						count += 1;
						offset = line_end;
				} else {
					break count;
				}
    }
```

Reply:
Yes, it is different but `count_lines_until` can be expressed by `count_lines(TextRange::up_to(offset))` 

That's why I'd prefer to only expose the more generic solution because it allows solving more use cacses.